### ISSUE TITLE
Poetry Torch 2.1.1+cu121 problem on Windows

### ISSUE THREAD

**Issue by MrMarvel** [pytorch/pytorch#114042]
### 🐛 Describe the bug

https://download.pytorch.org/whl/cu121/torch-2.1.1%2Bcu121-cp310-cp310-linux_x86_64.whl
On torch 2.1.1 linux package there are requirements for linux with no filter for platform which conflicts with poetry on Windows
![image](https://github.com/pytorch/pytorch/assets/24212110/8993bb1a-ac05-4587-a1fa-49f42052ebd8)
![image](https://github.com/pytorch/pytorch/assets/24212110/0bf3cfaa-6f7b-4dfd-8ef7-4b8dfab5ac44)
![image](https://github.com/pytorch/pytorch/assets/24212110/43313b04-9379-4a95-aebc-c1d8fd0fbd93)

Please format your METADATA for linux package.
### Versions

2.1.1+cu121

cc @seemethere @malfet @osalpekar @atalman @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite @ezyang @gchanan @zou3519 @kadeng

**Comment by MrMarvel** [comment]
Poetry contributers says that I need to raise problem there.
https://github.com/python-poetry/poetry/issues/8690

**Comment by malfet** [comment]
Interesting.. I don't think we make any guarantees about poetry for wheels that is published on https://download.pytorch.org, but perhaps we should

For the reference on pypi dependencies looks as follows:
```
% curl -s https://pypi.org/pypi/torch/2.1.1/json | jq '.info.requires_dist'
[
  "filelock",
  "typing-extensions",
  "sympy",
  "networkx",
  "jinja2",
  "fsspec",
  "nvidia-cuda-nvrtc-cu12 (==12.1.105) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "nvidia-cuda-runtime-cu12 (==12.1.105) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "nvidia-cuda-cupti-cu12 (==12.1.105) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "nvidia-cudnn-cu12 (==8.9.2.26) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "nvidia-cublas-cu12 (==12.1.3.1) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "nvidia-cufft-cu12 (==11.0.2.54) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "nvidia-curand-cu12 (==10.3.2.106) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "nvidia-cusolver-cu12 (==11.4.5.107) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "nvidia-cusparse-cu12 (==12.1.0.106) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "nvidia-nccl-cu12 (==2.18.1) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "nvidia-nvtx-cu12 (==12.1.105) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "triton (==2.1.0) ; platform_system == \"Linux\" and platform_machine == \"x86_64\"",
  "jinja2 ; extra == 'dynamo'",
  "opt-einsum (>=3.3) ; extra == 'opt-einsum'"
]
```

**Comment by malfet** [comment]
Removing high priority tag as it affects only Windows platform (MacOS and Linux work fine)

**Comment by malfet** [comment]
To clarify the problem here a bit: dependency for the wheels we publish to PyPI are correct, but ones we publish to say https://download.pytorch.org/whl/cu118/
contain Linux wheels that depend on triton for all platforms

**Comment by malfet** [comment]
@ozanMSFT just to test that it actually works before release.
One should be able to download release candidate from https://download.pytorch.org/whl/test/cu121

**Comment by dimbleby** [comment]
the various distributions still carry different metadata, providing further opportunities for confusion.  

I compared https://download.pytorch.org/whl/test/cu118/torch-2.1.2%2Bcu118-cp310-cp310-linux_x86_64.whl and https://download.pytorch.org/whl/test/cu118/torch-2.1.2%2Bcu118-cp310-cp310-win_amd64.whl

- the linux distribution says that it provides the extra 'dynamo' which requires jinja2, the windows distribution makes no mention of this
- the windows distribution has linux-and-x86-64-only requirements for lots of nvidia packages

probably this will mostly work out because the linux one is alphabetically first and so probably that's the one that eg poetry will look at; and I guess the bonus extra is mostly benign.

but eg if poetry happened to look at the windows wheel to derive requirements, it would infer that the linux distribution wanted all those nvidia packages: which I suppose it doesn't.

The goal should be for all of the distributions to carry the same information.

**Comment by malfet** [comment]
Update: poetry does not seem to be even try to parse metadata from anything but official pypi repo

**Comment by dimbleby** [comment]
> Update: poetry does not seem to be even try to parse metadata from anything but official pypi repo

this is certainly untrue eg per the poetry issue that was the reason this one was raised

**Comment by dimbleby** [comment]
> what issue you are referring to?

https://github.com/pytorch/pytorch/issues/114042#issuecomment-1817718386

PEP658 is not currently supported by poetry anywhere (including pypi) but it should be coming soon, hopefully - https://github.com/python-poetry/poetry/pull/5509 looks close to merging

you are mistaken about `data-requires-python`, see https://github.com/python-poetry/poetry/blob/0578ed81787449d79430360de143d916006e028b/src/poetry/repositories/link_sources/html.py#L37

**Comment by atalman** [comment]
I can confirm with cherry pick landed on 2.1.2
release: 2.1.1 METADATA linux (Large wheels)
```
Requires-Dist: triton (==2.1.0)
```
With  2.1.2 METADATA linux, Windows cu118 and cu121 (Large wheels)
```
Requires-Dist: triton (==2.1.0) ; platform_system == "Linux" and platform_machine == "x86_64"
```

**Comment by melMass** [comment]
@malfet sorry to revive an old issue but is there a way to solve that for versions prior to 2.1.2? 
I'm using a runtime that depends on 2.1.1 but can't find a way to properly install it because of this marker issue
